src/angular/tooltip/tooltip.directive.ts

```diff
@@ -0,0 +1,302 @@
+
+        this.destroy();
+    }
+
+    private toggleShowCssClass(isAdd: boolean) {
+        if (this.tooltip) {
+            this.renderer.setElementClass(this.tooltip, this.cssClass + '-' + showSuffix, isAdd);
+        }
+    }
+
+    /**
+     * Adds placement css class and sets tooltip position in style
+     */
+    private setPosition() {
+        const tooltipPos: IPlacementData = this.getPlacementData();
+
+        const placementSuffix: string = TooltipPlacement[tooltipPos.placement].toLowerCase();
+        this.renderer.setElementClass(this.tooltip, this.cssClass + '-' + placementSuffix, true);
+
+        this.renderer.setElementStyle(this.tooltip, topStyle, tooltipPos.top + pixel);
+        this.renderer.setElementStyle(this.tooltip, leftStyle, tooltipPos.left + pixel);
+    }
+
+    /**
+     * Checks the specified placement (first element in array), if it is not valid - checks other placements
+     * @returns {IPlacementData}
+     */
+    private getPlacementData(): IPlacementData {
+        const placement: TooltipPlacement = this.placement;
+        let tooltipPos: IPlacementData;
+
+        const tooltipPosWithPlacement = this.getPlacement.bind(this, placement);
+
+        // TODO add comments - done
+        switch (placement) {
+            case TooltipPlacement.Left:
+                tooltipPos = tooltipPosWithPlacement(
+                    TooltipPlacement.Right,
+                    TooltipPlacement.Top,
+                    TooltipPlacement.Bottom);
+                break;
+
+            case TooltipPlacement.Right:
+                tooltipPos = tooltipPosWithPlacement(
+                    TooltipPlacement.Left,
+                    TooltipPlacement.Top,
+                    TooltipPlacement.Bottom);
+                break;
+
+            case TooltipPlacement.Top:
+                tooltipPos = tooltipPosWithPlacement(
+                    TooltipPlacement.Bottom,
+                    TooltipPlacement.Left,
+                    TooltipPlacement.Right);
+                break;
+
+            case TooltipPlacement.Bottom:
+                tooltipPos = tooltipPosWithPlacement(
+                    TooltipPlacement.Top,
+                    TooltipPlacement.Left,
+                    TooltipPlacement.Right);
+                break;
+        }
+
+        return tooltipPos;
+    }
+
+    /**
+     * Returns valid tooltip position data
+     * @param {TooltipPlacement} placement
+     * @param {TooltipPlacement} additionalPlacements
+     * @returns {IPlacementData}
+     */
+    private getPlacement(placement: TooltipPlacement,
+                         ...additionalPlacements: TooltipPlacement[],
+                         ): IPlacementData {
+        const placements: TooltipPlacement[] = [placement, ...additionalPlacements];
+        const filterPlacements = placements
+            .map((pl) => this.getPosition(pl))
+            .filter((item) => this.validatePosition(item));
```

Comment:
Performance: You can combine `map` and `filter` to `find`, which will stop at the first valid position.